### PR TITLE
Add more to grammer

### PR DESCRIPTION
**Pull request by soelgary**
This is a work in progress in order to get the grammar for the language

### CODE REVIEW

File: src/parser/tiger.grm [hunk 1]
```diff
@@ -73,3 +77,22 @@ op: PLUS ()
...
     | GE ()
+
+arraycreateexp: ID LBRACK exp RBRACK OF exp ()
+
+recordcreateexp: ID LBRACE recordcreatelistexp RBRACE ()
+
+recordcreatelistexp: recordcreatelist ()
+    | ()
+
+recordassignment: ID EQ exp ()
```
Comment (nixpulvis) [inline src/parser/tiger.grm]: Following from the grammar, this should be `fieldcreate` right?

File: src/parser/tiger.grm [hunk 2]
```diff
@@ -73,3 +77,22 @@ op: PLUS ()
     | LE ()
     | GT ()
     | GE ()
+
+arraycreateexp: ID LBRACK exp RBRACK OF exp ()
+
+recordcreateexp: ID LBRACE recordcreatelistexp RBRACE ()
+
+recordcreatelistexp: recordcreatelist ()
+    | ()
```
Comment (nixpulvis) [inline src/parser/tiger.grm:99]: I wouldn't make an extra state for this. Also this should look like the code we had for `callexp` which I believe just from reading this is buggy, as it should allow `foo()` which it doesn't atm. Correct?

Comment (nixpulvis) [inline src/parser/tiger.grm:99]: General form might be

```
something: LDILIM somethinglist RDILIM ()  // This is the one or more state.
         | LDILIM RDILIM ()                // This is the empty state.

somethinglist: somethinglist LISTDILIM inner ()
             | inner ()
```

File: src/parser/tiger.grm (hunk 1 above)
Comment (soelgary) [inline src/parser/tiger.grm]: Yea thats a bad name

File: src/parser/tiger.grm (hunk 2 above)
Comment (soelgary) [inline src/parser/tiger.grm:99]: this is fixed in #21

Comment (nixpulvis) [inline src/parser/tiger.grm:99]: Can you make this match?

Comment (soelgary) [inline src/parser/tiger.grm:99]: yep

File: src/parser/tiger.grm [hunk 3]
```diff
@@ -40,37 +48,72 @@
...
+assignmentexp: lvalue ASSIGN exp ()
+
+lvalue: ID ()
+    | lvalue DOT ID ()
+    | lvalue LBRACK exp RBRACK ()
+    | ID LBRACK exp RBRACK ()
+
+ifthenelseexp: IF exp THEN exp ELSE exp ()
+
+ifthenexp: IF exp THEN exp ()
```
Comment (nixpulvis) [inline src/parser/tiger.grm:115]: How'd you get around the ambiguity here?

File: .gitignore [hunk 4]
```diff
@@ -4,3 +4,6 @@
 tiger/
 
 !sources.cm
+
+*.grm.desc
+*.grm.sml
```
Comment (nixpulvis) [inline .gitignore:9]: Might also want to ignore `*.grm.sig` while your at it.

File: src/parser/tiger.grm (hunk 3 above)
Comment (soelgary) [inline src/parser/tiger.grm:115]: I made the precedence of `else` higher than `then` by doing this.

``` text
%nonassoc THEN
%nonassoc ELSE
```

This makes each `else` bind to the nearest `if`(which I believe is what we need to do)

Comment (nixpulvis) [inline src/parser/tiger.grm:115]: I'm pretty sure this just makes it a syntax error. From what I'm reading anyway.

File: .gitignore (hunk 4 above)
Comment (soelgary) [inline .gitignore:9]: Is that autogenerated?

Comment (nixpulvis) [inline .gitignore:9]: Yessir

**Pull request merged**